a3p-integration/proposals/d:liquidation-visibility/post.liquidation.js

```diff
@@ -0,0 +1,289 @@
+ * - agoric run on both of them
+ */
+test.serial('prepare vault factory', async t => {
+  // Prepare mutated vaultFactory
+  const rootRW = makeFileRW('.', { fsp, path });
+  const termsW = rootRW.join('./termsWrapper.js');
+  const termsR = termsW.readOnly();
+  const content = await termsR.readText();
+  console.log('TERM WRAPPER');
+  console.log(content);
+
+  const replaceText = 'termsWrapper(zcf.getTerms(), privateArgs);';
+
+  // Make vaultManager use the custom timer
+  const vmRW = rootRW.join('./artifacts/src/vaultFactory/vaultManager.js');
+  const vmVersion2 = rootRW.join(
+    './artifacts/src/vaultFactory/vaultManagerV2.js',
+  );
+  const vmRead = vmRW.readOnly();
+  const vmText = await vmRead.readText();
+  const vmMutated = vmText
+    .replace(
+      '{ zcf, marshaller, makeRecorderKit,' + ' factoryPowers }',
+      '{ zcf, marshaller, makeRecorderKit, factoryPowers,' + ' timerService }',
+    )
+    .replace(
+      '{ priceAuthority, timerService,' + ' reservePublicFacet' + ' }',
+      '{ priceAuthority, reservePublicFacet }',
+    );
+  await vmVersion2.writeText(vmMutated);
+
+  // Make vaultDirector forward the custom timer
+  const vdRW = rootRW.join('./artifacts/src/vaultFactory/vaultDirector.js');
+  const vdVersion2 = rootRW.join(
+    './artifacts/src/vaultFactory/vaultDirectorV2.js',
+  );
+  const vdRead = vdRW.readOnly();
+  const vdText = await vdRead.readText();
+  const vdMutated = vdText
+    .replace('./vaultManager', './vaultManagerV2')
+    .replace(
+      '\n' +
+        '    makeERecorderKit,\n' +
+        '    makeRecorderKit,\n' +
+        '    marshaller,\n' +
+        '    factoryPowers,\n' +
+        '    zcf,\n' +
+        '  }',
+      '\n' +
+        '    makeERecorderKit,\n' +
+        '    makeRecorderKit,\n' +
+        '    marshaller,\n' +
+        '    factoryPowers,\n' +
+        '    zcf,\n' +
+        '    timerService: timer,\n' +
+        '  }',
+    )
+    .replace(
+      'fn(vm)',
+      "Promise.resolve(vm).then(vm => fn(vm)).catch(e => trace('ERROR: allManagersDo', e))",
+    );
+  await vdVersion2.writeText(vdMutated);
+
+  // Override vaultFactory zcf.getTerms();
+  const vfRW = rootRW.join('./artifacts/src/vaultFactory/vaultFactory.js');
+  const vfVersion2 = rootRW.join(
+    './artifacts/src/vaultFactory/vaultFactoryV2.js',
+  );
+  const vfRead = vfRW.readOnly();
+  const vfText = await vfRead.readText();
+  const vfMutatedTermsOnly = vfText.replace('zcf.getTerms();', replaceText);
+  const vfMutatedFinal = vfMutatedTermsOnly.replace(
+    './vaultDirector',
+    './vaultDirectorV2',
+  );
+  await vfVersion2.writeText(content + '\n' + vfMutatedFinal);
+
+  t.pass();
+});
+
+test.serial('build proposal', async t => {
+  await copyAll(
+    [
+      {
+        src: './artifacts/src/vaultFactory/vaultFactoryV2.js',
+        dest: '/usr/src/agoric-sdk/packages/inter-protocol/src/vaultFactory/vaultFactoryV2.js',
```

Comment:
there should never be a `vaultFactoryV2.js` in the SDK paths. There is simply `vaultFactory.js` which is whatever the HEAD version is. The earlier version is on-chain.

Reply:
I understand where you are coming from. The reason I felt comfortable doing what you've highlighted is that `vaultFactoryV2.js` never **persists** to the next build as we run `post.liquidation.js` in the `TEST` phase of the a3p process. See [here](https://github.com/Agoric/agoric-sdk/pull/8994/files#diff-2cc252819b9faa44934f1f43ac67cdde53c9fd83714bf6be527570f35063284c).

The reason why we make changes to the HEAD is, we are trying to trigger a liquidation. In order to test the things that depend on a liquidation we need to control the liquidation process. This means we need to have a `timerService` we can control which is used by both `auctioneer` and `vaultFactory`. 

Triggering and controlling a liquidation is crucial for us as we try to test the state **after** a liquidation. I hope this justifies having a `vaultFactoryV2.js` in the SDK path in non-persisting `TEST` phase.

Reply:
If it never persists, why put it under `agoric-sdk/packages/inter-protocol`? Can't you build it from somewhere in this proposal path?

Reply:
Building `vaultFactory` outside of the sdk path seemed a little tricky at the time given all the dependencies and all that. Maybe I can copy `interProtocol` in the proposal path then build it there 🤔

Reply:
Ah, true. So you need a special version of `vaultFactory` that's just for testing purposes? We have`zcf.setTestJig` for in-place testing logic, but a3p tests are for designed for production fidelity and so can't reach into that. 

If this contract is just for testing, and not a hypothetical upgrade, then I don't see a problem with it being under agoric-sdk but with a name like 'vaultFactoryForTesting.js`. In which case I even think we should consider having it in master so it can be used again in the future and it doesn't stagnate from regular VF.

Reply:
> If this contract is just for testing, and not a hypothetical upgrade, then I don't see a problem with it being under agoric-sdk but with a name like 'vaultFactoryForTesting.js`. In which case I even think we should consider having it in master so it can be used again in the future and it doesn't stagnate from regular VF.

Yeah I can totally agree with that and believe this is to be the case for us. However, this stagnation problem keeps me from moving forward with creating a liquidation in `a3p` as `auctioneer` is also a part of this liquidation. The best solution I can think of is;
 * Put `manualTimer` inside a contract and deploy it
   * `creatorFacet` will have methods for advancing time
   * `publicFacet` will hand out the reference to the `manualTimer`
 * Spin up a new instance of `auctioneer` using the original auctioneer's installation(assuming auctioneer is not upgradable)
   * Have the  fakeAuctioneer use `manualTimer` as its timer service
 * Upgrade `vaultFactory` by using `vaultFactoryForTesting.js` as its entry point when bundling which uses `manualTimer` and `fakeAuctioneer`
   * The thing is we need `vaultFactoryForTesting.js` to have `vaultFactory` accept its timer service from `privateArgs`. I see that `vaultFactory` recently moved `auctioneerPF` to its private args. If we do the same for timer service, we won't need `vaultFactoryForTesting.js` at all.

I am curious about any suggestions you might have as well as what you think about this strategy. @turadg